### PR TITLE
Attempting to fix bad highlighting

### PR DESCRIPTION
**Pull request by alexmojaki**
By @aroberge

### CODE REVIEW

**Comment by aroberge** [comment]
I am guessing that the failing test is due to https://github.com/alexmojaki/stack_data/blob/master/tests/golden_files/f_string_new.txt not being reproduced. Line 58 (as indicated) has some leading spaces highlighted, which would be (correctly) removed by my changes.

File: stack_data/core.py [hunk 1]
```diff
@@ -678,7 +676,10 @@ def _executing_node_common_indent(self) -> int:
                 indent = max(rang.start, begin_text)
                 indents.append(indent)
 
-        return min(indents) if indents else 0
+        if len(indents) <= 1:
+            return 0
+
+        return min(indents[1:])
```
Comment (alexmojaki) [inline stack_data/core.py:682]: I don't really understand what the goal of this new approach is, but this line in particular looks like it's assuming the the current line is the first highlighted line. Remember that isn't always the case https://github.com/alexmojaki/stack_data/pull/20#discussion_r803015541

**Comment by alexmojaki** [comment]
Can you describe in plain English what the new method does in general? It's not clear to me what you're going for here and what it might change in cases that aren't in the tests.

I'm not actually sure that anything needs to change. For example I'm not really sure that this:

```
-->    5 | nb = len(letter
                ^^^^^^^^^^
       6 |          for letter in "word")
                ^^^^^^^^^^^^^^^^^^^^^^^^^
```

is actually worse than this:

```
-->    5 | nb = len(letter
                ^^^^^^^^^^
       6 |          for letter in "word")
                    ^^^^^^^^^^^^^^^^^^^^^
```

I know I said it doesn't look great, but I'm having second thoughts now.

File: stack_data/core.py (hunk 1 above)
Comment (aroberge) [reply stack_data/core.py:682]: No, this line is assuming that the first highlighted line might be highlighted with a different leading indentation than all the lines below it. 
With Python 3.9:
```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 5, in <module>
       1 | import stack_data
       3 | stack_data.Formatter().set_hook()
-->    5 | nb = len(letter
                ^^^^^^^^^^
       6 |          for letter
                    ^^^^^^^^^^
       7 |          in "word")
                    ^^^^^^^^^^
TypeError: object of type 'generator' has no len()
```
With Python 3.6:
```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 7, in <module>
       1 | import stack_data
       3 | stack_data.Formatter().set_hook()
       5 | nb = len(letter
                ^^^^^^^^^^
       6 |          for letter
                    ^^^^^^^^^^
-->    7 |          in "word")
                    ^^^^^^^^^^
TypeError: object of type 'generator' has no len()
```

**Comment by aroberge** [comment]
Ok, here's an attempt to explain it. If you recall the "old way" of highlighting, which I modified, in https://github.com/alexmojaki/stack_data/pull/20#discussion_r803015541, when there were more than one line highlighted, the lines below the first one were highlighted starting at column zero.  This meant that leading spaces were highlighted. This didn't look particularly good, especially when Pygment was used to highlight the code.

The approach I used previously was to consider **all** the highlighted lines, and look for the common indentation required to capture all the highlighted block of code, minimizing the amount of leading space highlighted. Once that common indentation was calculated, the highlighting took place, usually with all the lines highlighted starting at the first column, but sometimes with an exception for the first highlighted line if the indentation started at a column beyond the "common indentation". Here's an extreme example (using the new code, which gives the same result for this case):
```
       1 | import stack_data
       3 | stack_data.Formatter().set_hook()
       5 | nb_characters = len(letter
                           ^^^^^^^^^^
       6 |          for letter
                    ^^^^^^^^^^
       7 |            in
                    ^^^^
-->    8 |          "word")
                    ^^^^^^^
```
To calculate the "common minimal indentation" of the "block" below the first highlighted line, the indentation of the first line was taken into account. In the case above, the column where the highlighting starts for the first line is greater than the common minimal indentation -- so, it does not affect the value of the common minimal indentation.

The new code explicitly excludes where the highlighting of the first line starts from the computation of the common minimal indentation. So, in the example below, the highlighted "block" is identical to the example above, even though the highlighting of the first line starts at an earlier column:
```
       1 | import stack_data
       3 | stack_data.Formatter().set_hook()
       5 | nb = len(letter
                ^^^^^^^^^^
       6 |          for letter
                    ^^^^^^^^^^
       7 |            in
                    ^^^^
-->    8 |          "word")
                    ^^^^^^^
```

**Comment by alexmojaki** [comment]
OK, I think I understand a bit better now. I'm still not sure if this is necessarily an improvement, or if it makes sense to treat the first line as special in this way.

**Comment by alexmojaki** [comment]
Unfortunately this stuff doesn't affect highlighting with Pygments. If you try these cases on https://futurecoder.io/course/#ide or in IPython you'll see that leading indentation is still highlighted.

**Comment by aroberge** [comment]
Here's what friendly currently does (which is admittedly not great in this contrived case):
image. When highlighting with background colour, I do additional processing which ends up removing leading spaces highlighted where it should not. I'll need to fix this.

Here is a more realistic case, in terms of code indentation:
image

**Comment by alexmojaki** [comment]
OK, let's go forward with this. Can you add a couple of tests specifically for this? This one is a good example:

```python
       1 | import stack_data
       3 | stack_data.Formatter().set_hook()
       5 | nb = len(letter
                ^^^^^^^^^^
       6 |          for letter
                    ^^^^^^^^^^
       7 |            in
                    ^^^^
-->    8 |          "word")
                    ^^^^^^^
```

**Comment by aroberge** [comment]
Yes I will add some extra tests. I'll try to this later today.

**Comment by aroberge** [comment]
I've added two tests: one where the highlighted block starts on a column left to that of the highlight on the first line, and one where it starts to the right.

**Comment by alexmojaki** [comment]
Thanks!
